Draw peak marker using cached position if no detector

jclarkeSTFC opened the pull request:
Peaks have a cached position, but they may not have a detector, so we can't just rely on the detector ID to draw the marker.

Fixes #37514 

Peaks have a cached position, which is where the position is stored when there's no corresponding detector. This can happen because it is possible to have peaks beyond the bounds of any detector, which is what is occurring the files in the linked issue. I added a method to the `UnwrappedSurface` base class to project a position onto the surface, so in `PeakOverlay::createMarkers()` we can use that method combined with the cached peak position. It's not possible to do this with a `PanelsSurface`, the code requires a detector, but in that case we shouldn't have a detectorless peak in the first place, plus there's a try-catch there just in case. Previously if this had happened it would have caused a sad mantid anyway.

### To test:
See #37514 for how to reproduce.

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
If no detector ID, then draw marker from position

Peaks have a cached position, but they may not have a detector, so we
can't just rely on the detector ID to draw the marker.

jclarkeSTFC committed:
Add release notes

jclarkeSTFC committed:
Mark arguments as unused to fix compiler warnings

jclarkeSTFC committed:
Add override in test

jclarkeSTFC committed:
Add test for overlaying a peak with no detector

jhaigh0 commented on qt/widgets/instrumentview/test/InstrumentWidget/InstrumentWidgetTest.h:
```diff
@@ -299,6 +314,30 @@ class InstrumentWidgetTest : public CxxTest::TestSuite {
+      auto peaksWs = std::dynamic_pointer_cast<IPeaksWorkspace>(ws);
+      auto peak = Mantid::DataObjects::Peak(peaksWs->getInstrument(), Mantid::Kernel::V3D(1, 1, 1));
+      peaksWs->addPeak(peak);
+      widget.overlay("peaks");
```
up to you but I like to add `TS_ASSERT_THROWS_NOTHING` around the line where I'm testing no crash occurs just to make it explicit what the point of the test is. I do understand it's pretty redundant though :upside_down_face:

jhaigh0 commented on qt/widgets/instrumentview/test/InstrumentWidget/InstrumentWidgetTest.h:
```diff
@@ -6,12 +6,14 @@
 #include "MantidQtWidgets/InstrumentView/InstrumentWidgetMaskTab.h"
 
 #include "MockGLDisplay.h"
+#include "MockInstrumentActor.h"
```
not sure you need this include

jhaigh0 reviewed (commented):
Looks good, was able to overlay the peaks with no crash.

RichardWaiteSTFC commented:
Just checked this and works well on cylindrical and spherical projections. For side-by-side view it didn't crash and just threw an error
`A detector ID is required to project with a PanelsSurface.`
which is good - though it may be helpful to print the peak index in the table? 
I was just wondering where this cached position comes from, the other option is to just not plot it on all projections?

jclarkeSTFC committed:
Add more info to error when overlaying peaks

jclarkeSTFC commented:
> Just checked this and works well on cylindrical and spherical projections. For side-by-side view it didn't crash and just threw an error `A detector ID is required to project with a PanelsSurface.` which is good - though it may be helpful to print the peak index in the table? I was just wondering where this cached position comes from, the other option is to just not plot it on all projections?

I've added the peak index and the detector ID to the error, the force push is because I rebased on top of the Boost update.

It looks like the cached position is either from the detector, e.g. `Peak.cpp::216`, or is from `Peak::getVirtualDetectorPosition()`, which is called inside `setQLabFrame` and in a constructor that uses a scattering angle to calculate where the detector would be.

Yes I guess the other option would be to just skip peaks with no detector ID when creating markers, could log an error for those peaks to let the user know what's going on.

RichardWaiteSTFC commented:
> > Just checked this and works well on cylindrical and spherical projections. For side-by-side view it didn't crash and just threw an error `A detector ID is required to project with a PanelsSurface.` which is good - though it may be helpful to print the peak index in the table? I was just wondering where this cached position comes from, the other option is to just not plot it on all projections?
> 
> I've added the peak index and the detector ID to the error, the force push is because I rebased on top of the Boost update.
> 
> It looks like the cached position is either from the detector, e.g. `Peak.cpp::216`, or is from `Peak::getVirtualDetectorPosition()`, which is called inside `setQLabFrame` and in a constructor that uses a scattering angle to calculate where the detector would be.
> 
> Yes I guess the other option would be to just skip peaks with no detector ID when creating markers, could log an error for those peaks to let the user know what's going on.

Thanks for adding the info to the error, I think if the virtual detector position is sensible (based on spherical angles or cylindrical angles to kf) it could be really useful as it lets the user know where the peak is, how they can rotate the sample to bring it on the detector. But if the position is nonsense then it could in fact be mis-leading!
